File: backend/app/schemas/currency.py
"""Supported recorded units. Normalization does not convert monetary values."""
from typing import Annotated
from pydantic import BeforeValidator

SUPPORTED_CURRENCIES = {"PHP", "USD", "NZD", "AUD", "EUR", "GBP", "CAD"}


def normalize_currency(value):
    code = value.strip().upper() if isinstance(value, str) else ""
    return code if code in SUPPORTED_CURRENCIES else None


def validate_currency(value):
    code = normalize_currency(value)
    if code is None:
        raise ValueError("Choose a supported currency: PHP, USD, NZD, AUD, EUR, GBP or CAD")
    return code


Currency = Annotated[str, BeforeValidator(validate_currency)]

File: frontend/lib/currency.ts
// Recorded units only: normalization never converts an amount.
export const currencies = ["PHP", "USD", "NZD", "AUD", "EUR", "GBP", "CAD"] as const;
export function normalizeCurrency(value: unknown): string | null {
  const code = typeof value === "string" ? value.trim().toUpperCase() : "";
  return currencies.some(currency => currency === code) ? code : null;
}
export function planningCurrency(country: string): string | null {
  return ({ Philippines: "PHP", "New Zealand": "NZD", Australia: "AUD", "United States": "USD" } as Record<string, string>)[country] ?? null;
}
export function normalizeTicker(value: unknown): string {
  return typeof value === "string" ? value.trim().toUpperCase() : "";
}
